tq/ssh.go

```diff
@@ -0,0 +1,405 @@
+package tq
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/git-lfs/git-lfs/errors"
+	"github.com/git-lfs/git-lfs/lfshttp"
+	"github.com/git-lfs/git-lfs/ssh"
+	"github.com/git-lfs/git-lfs/tools"
+	"github.com/rubyist/tracerx"
+)
+
+type SSHBatchClient struct {
+	maxRetries int
+	transfer   *ssh.SSHTransfer
+}
+
+func (a *SSHBatchClient) batchInternal(args []string, batchLines []string) (int, []string, []string, error) {
+	conn := a.transfer.Connection(0)
+	conn.Lock()
+	defer conn.Unlock()
+	err := conn.SendMessageWithLines("batch", args, batchLines)
+	if err != nil {
+		return 0, nil, nil, errors.Wrap(err, "batch request")
+	}
+
+	status, args, lines, err := conn.ReadStatusWithLines()
+	if err != nil {
+		return status, nil, nil, errors.Wrap(err, "batch response")
+	}
+	return status, args, lines, err
+}
+
+func (a *SSHBatchClient) Batch(remote string, bReq *batchRequest) (*BatchResponse, error) {
+	bRes := &BatchResponse{TransferAdapterName: "ssh"}
+	if len(bReq.Objects) == 0 {
+		return bRes, nil
+	}
+
+	missing := make(map[string]bool)
+	batchLines := make([]string, 0, len(bReq.Objects))
+	for _, obj := range bReq.Objects {
+		missing[obj.Oid] = obj.Missing
+		batchLines = append(batchLines, fmt.Sprintf("%s %d", obj.Oid, obj.Size))
+	}
+
+	tracerx.Printf("api: batch %d files", len(bReq.Objects))
+
+	requestedAt := time.Now()
+	args := []string{"transfer=ssh"}
+	if bReq.Ref != nil {
+		args = append(args, fmt.Sprintf("refname=%s", bReq.Ref.Name))
+	}
+	status, args, lines, err := a.batchInternal(args, batchLines)
```

Comment:
I believe the `args` return value is ignored after this, so this could be either:
```go
 	status, _, lines, err := a.batchInternal(args, batchLines)
```
 or `batchInternal()` could not even bother to return `args` at all, so:
```go
 	status, lines, err := a.batchInternal(args, batchLines)
```

If the latter case is preferred, then I'd further suggest that in `batchInternal()` we change to:
```go
 	status, _, lines, err := conn.ReadStatusWithLines()`
```

Reply:
I've just dropped the extra args in both places.

Reply:
Awesome, thank you.  I think there's one last tweak possible here, which would be to use `status, _, lines, err := conn.ReadStatusWithLines()` in `batchInternal()`, just to make it clear we're ignoring the `args` return value.

Reply:
Great, I've just fixed that.